python/lsst/meas/base/forcedPhotCcd.py

```diff
@@ -186,6 +199,90 @@ class ForcedPhotCcdTask(ForcedPhotImageTask):
+        """Filter reference catalog so that all sources are within the
+        boundaries of the exposure.
+
+        Parameters
+        ----------
+        exposure : `lsst.afw.image.exposure.Exposure`
+            Exposure to generate the catalog for.
+        refCat : `lsst.afw.table.SourceCatalog`
+            Catalog of shapes and positions at which to force photometry.
+        refWcs : `lsst.afw.image.SkyWcs`
+            Reference world coordinate system.
+
+        Returns
+        -------
+        sources : `lsst.afw.table.SourceCatalog`
+            Filtered catalog of forced sources to measure.
+
+        Notes
+        -----
+        Filtering the reference catalog is currently handled by Gen2
+        specific methods.  To function for Gen3, this method copies
+        code segments to do the filtering and transformation.  The
+        majority of this code is based on the methods of
+        lsst.meas.algorithms.loadReferenceObjects.ReferenceObjectLoader
+
+        """
+
+        # Step 1: Determine bounds of the exposure photometry will
+        # be performed on.
+        expWcs = exposure.getWcs()
+        expRegion = exposure.getBBox(lsst.afw.image.PARENT)
+        expBBox = lsst.geom.Box2D(expRegion)
+        expBoxCorners = expBBox.getCorners()
+        expSkyCorners = [expWcs.pixelToSky(corner).getVector() for
+                         corner in expBoxCorners]
+        expPolygon = lsst.sphgeom.ConvexPolygon(expSkyCorners)
+
+        # Step 2: Filter out reference catalog sources that are
+        # not contained within the exposure boundaries.
+        sources = type(refCat)(refCat.table)
+        for record in refCat:
+            if expPolygon.contains(record.getCoord().getVector()):
+                sources.append(record)
+        refCatIdDict = {ref.getId(): ref.getParent() for ref in sources}
+
+        # Step 3: Cull sources that do not have their parent
+        # source in the filtered catalog.  Save two copies of each
+        # source.
+        sources = type(refCat)(refCat.table)
+        refSources = type(refCat)(refCat.table)
+        for record in refCat:
+            if expPolygon.contains(record.getCoord().getVector()):
+                recordId = record.getId()
+                topId = recordId
+                while (topId > 0):
+                    if topId in refCatIdDict:
+                        topId = refCatIdDict[topId]
+                    else:
+                        break
+                if topId == 0:
+                    sources.append(record)
+                    refSources.append(record)
+
+        # Step 4: Transform source footprints from the reference
+        # coordinates to the exposure coordinates.
+        for srcRecord, refRecord in zip(sources, refSources):
+            srcRecord.setFootprint(refRecord.getFootprint().transform(refWcs,
+                                                                      expWcs, expRegion))
```

Comment:
I don't see why we need two catalogs containing the same records; couldn't you just do:
```
for refRecord in refSources:
    refRecord.setFootprint(refRecord.getFootprint().transform(refWcs, expWcs, expRegion))
```
?